### DISCUSSION TITLE
How to add terrainshader / layershader via C++

### DISCUSSION THREAD

**Discussion by christ0phd** [pelicanmapping/osgearth#2136] [Support Q&A] [answered]
Hi all,

I'm trying to integrate this [shader script](https://gist.github.com/gwaldron/e60c7ce2b0eb1ebb7418249caaebb5ba) into my C++ source to have a vertical scaling in my terrain. But without any luck so far.
Placing the shader script into an earth file is working.
But what is the intended way of integrating the shader script into the osgEarth map? I tried adding it as a virtual programm and as a layer shader. Both with no result.
Is there any howto or any guidance on this topic?

Christoph

**Comment by hswjb** [comment]
maybe osgearth/contourmap?

**Comment by RhythmOfTension** [comment, marked as answer]
I think you can get some inspiration for `TerrainEffect` creation from osgEarthUtil/VerticalScale in 2.10 branch.  And then add it to `TerrainEngine` via `addEffect`. But be aware of some problems with vertical scaling, see for example: https://github.com/gwaldron/osgearth/issues/2142

**Comment by horrzs** [comment]
Do you have any new developments？ @christ0phd 
I have also been trying to meet this requirement recently
